src/basic/unit-file.c

```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
         return false;
 }
 
+static int unit_file_resolve_symlink(
+                const char *root_dir,
+                char **search_path,
+                const char *dir,
+                int dirfd,
+                const char *filename,
+                char **ret_simplified,
+                const char **ret_dst) {
```

Comment:
the semantics of ret_dst are so mean, i.e. returning either a pointer to the pointer or to the simplified allocated buffer.

Afaics the caller doesn't bother about the `simplified` return parameter at all. Only about `dst`. Which it eventually duplicates to add to the set. It sounds much simpler if we'd just drop ret_simplified here, and return ret_dst as properly allocated string, that the caller can then just add to the set as-is, without eeding to duplicate it.

It's simpler and faster...

Reply:
It's reworked along the suggested lines now.